core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -337,7 +337,15 @@ class ZkPartitionStateMachine(config: KafkaConfig,
 
       finished.foreach {
         case (partition, Left(e)) =>
-          logFailedStateChange(partition, partitionState(partition), OnlinePartition, e)
+          // Extract failure message, if present, to append to error log.
+          val reasonMsg = if (e.isInstanceOf[StateChangeFailedException]) {
+            s", reason: ${e.getMessage}"
+          } else {
+            ""
+          }
+          logger.error(s"Controller $controllerId epoch ${controllerContext.epoch} failed to "
+            + s" change state for partition $partition from ${partitionState(partition)} "
+            + s"to $OnlinePartition $reasonMsg", e)
```

Comment:
With this change we'll actually see the exception message twice since we are still including it in the call to `error`. Would it be worthwhile to check for `StateChangeFailedException` separately? Perhaps we can print the stack trace only when it is an unexpected exception type.

Reply:
Added separate cases for `StateChangeFailedException` and others.